Assets/Scripts/Bosses/MechBossAI.cs

```diff
@@ -0,0 +1,125 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class MechBossAI : MonoBehaviour
+{
+    public int currentHealth;
+
+    [SerializeField] Animator animator;
+    [SerializeField] MovementOverride movement;
+    [SerializeField] PlayerMovementController player;
+    [SerializeField] PlayerCombat playerCombat;
+    [SerializeField] Rigidbody2D rb;
+    [SerializeField] Resource health;
+    [SerializeField] AttackData stompData;
+    [SerializeField] AttackData missileData;
+
+    [SerializeField] float detectionRadius;
+    [SerializeField] float meleeRange;
+
+    Vector2 velocity;
+    bool isFacingRight = true;
+    float horizontalMove = 0.0f;
+    float maxSpeedChange;
+    float nextMeleeTime = 0f;
+    float nextProjectileTime = 0f;
+
+    // Values used for animations:
+
+    [SerializeField] bool isFiringProjectile = false;
+    [SerializeField] bool isStomping = false;
+    [SerializeField] GameObject cannon0;
+    [SerializeField] GameObject cannon1;
+
+    void Update()
+    {
+        float distanceToPlayer = Mathf.Abs(player.transform.position.x - transform.position.x);
```

Comment:
we probably want to null-check here to make sure the `player` is present - my usual approach for this sort of thing is to poll by some interval to look for the player if it is null like so:

```
IEnumerator FindingPlayer() {
  while (true) {
    player = FindPlayer();
    yield return new WaitForSeconds(<POLL_INTERVAL>);
  }
}

PlayerMovementController FindPlayer() {
  if (player != null) return player;
  GameObject playerObj = GameObject.FindWithTag(Constants.PLAYER_TAG);
  if (playerObj == null) return null;
  return playerObj.GetComponent<PlayerMovementController>();
}
```

Reply:
e.g. if the player dies, this will start throwing errors since we destroy the existing player object